stumpy/scrump.py

```diff
@@ -116,8 +116,11 @@ def _compute_PI(
 
         nn_idx = np.argmin(squared_distance_profile)
         if nn_idx not in I[thread_idx, i]:
-            # It is more than likely that the top-k values for the `i`-th subsequence
-            # will be already populated. So, we must shift-insert here
+            # Since the top-k values for the `i`-th subsequence may already
+            # be updated/populated in other previous iterations (i.e., not all
+            # values in `I[thread_idx]` are equal to `-1` or not all values in
+            # `P_squared[thread_idx, i]` are equal  to `np.inf`), we must
+            # shift-insert here rather than assign values to the first element.
             core._shift_insert_at_index(
```

Comment:
The approximate matrix profile **must** be sorted ascendingly. I don't think I am understanding your point here.

Reply:
So, please allow me to elaborate. So, as you can see in the code, we do:
```
# in the iteration that corresponds to `seq i`
nn_idx = np.argmin(dist_profile) # dist_profile of `seq i`
```
Then, if ` nn_idx not in I[thread_idx, i]`, we insert `nn_idx` at index 0 of its top-k profile `I[thread_idx, i]` (and update `P[thread_idx, i]` as well). This is in contrast to other parts of the `precsrump` function where we first perform `idx = np.searchsorted(top-k-profile, dist-to-neighbor)` to find the insertion index. And, the reason that we do not do `np.searchsorted(...)` for nn_idx is that we have the nearEST neighbor and thus it should have the smallest distance to `seq i` (and hence `nn_idx` is inserted at index 0). well...I think that is correct in theory! I mean, there might exist a case where the distance between `seq i` and one of its neighbors `seq j`  was already calculated in previous iterations and the value is different than `dist_profile[j]` due to imprecision. And, it might be even lower than the `dist_profile(nn_idx)` as well due to imprecision!!!


So, to make a long story short...I think we should better do `np.searchsorted` instead of just inserting `nn_idx` at index 0. We may do:
``` 
nn_idx = np.argmin(squared_distance_profile)
nn_dist = squared_distance_profile[nn_idx]
        if nn_idx not in I[thread_idx, i] and nn_dist <  I[thread_idx, -1]: # NEW CONDITION ADDED
            idx = np.searchsorted(P[thread_idx, i], nn_dist, side='right')  # NEW LINE
            # I think idx is 0 in 99.99% of the time.
            core._shift_insert_at_index(...)
            core._shift_insert_at_index(...)
```

Does that make sense? Please let me know if I should elaborate further.

Reply:
Cool, thanks for explaining and I can agree with your proposal that we can/should just do `np.searchsorted`. I like how this code is consistent and clear/obvious compared with other parts of the code

Reply:
> I really like how this code is consistent and clear/obvious in relation to other parts of the code. Let's do it!

Yeah..I remember it bothered you before that we inserted at index 0 without performing `np.searchsort()`. Now, the code becomes more consistent.

So, I am going to take care of these comments (and a few other things), and I will push again. This `prescrump`/`scrump` module was really tricky!

Reply:
Thank you for your persistence! I think that a lot of this work will be foundationally valuable and important that we get it as "right" as possible. I hope that you're not getting too burned out by it! Essentially, we're taking the original Matlab code and making it really robust (in addition to adding top-k). You're gaining experience in taking academic code and converting it into "production" usable code that is scalable and, hopefully, easy to maintain.

Reply:
> and important that we get it as "right" as possible. I hope that you're not getting too burned out by it!

Not at all :) I prefer to take care of these issues right now rather than leaving it for later. It is not easy for sure...but, it is fun :) and it helps me become more patient and improve my skills.

> You're gaining experience in taking academic code and converting it into "production" usable code that is scalable and, hopefully, easy to maintain.

Yeah...I listened to one of your interviews in a podcast in which you talked about this matter :) Resolving the issues of the (MATLAB) code, cleaning up the code, and optimizing it can make it easily reusable by others and open many doors for future developments of the algorithms.